keras_cv/datasets/pascal_voc/segmentation.py

```diff
@@ -0,0 +1,337 @@
+
+The image classification and object detection (bounding box) data is covered by existing
+TF datasets in https://www.tensorflow.org/datasets/catalog/voc. The segmentation data (
+both class segmentation and instance segmentation) are included in the VOC 2012, but not
+offered by TF-DS yet. This module is trying to fill this gap while TFDS team can
+address this feature (b/252870855, https://github.com/tensorflow/datasets/issues/27 and
+https://github.com/tensorflow/datasets/pull/1198).
+
+The schema design is similar to the existing design of TFDS, but trimmed to fit the need
+of Keras CV models.
+
+This module contains following functionalities:
+
+1. Download and unpack original data from Pascal VOC.
+2. Reprocess and build up dataset that include image, class label, object bounding boxes,
+   class and instance segmentation masks.
+3. Produce tfrecords from the dataset.
+4. Load existing tfrecords from result in 3.
+"""
+import logging
+import multiprocessing
+import os.path
+import tarfile
+import xml
+
+import tensorflow as tf
+
+DATA_URL = "http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCtrainval_11-May-2012.tar"
+
+# Note that this list doesn't contain the background class. In the classification use
+# case, the label is 0 based (aeroplane -> 0), whereas in segmentation use case, the 0 is
+# reserved for background, so aeroplane maps to 1.
+CLASSES = [
+    "aeroplane",
+    "bicycle",
+    "bird",
+    "boat",
+    "bottle",
+    "bus",
+    "car",
+    "cat",
+    "chair",
+    "cow",
+    "diningtable",
+    "dog",
+    "horse",
+    "motorbike",
+    "person",
+    "pottedplant",
+    "sheep",
+    "sofa",
+    "train",
+    "tvmonitor",
+]
+# This is used to map between string class to index.
+CLASS_TO_INDEX = {name: index for index, name in enumerate(CLASSES)}
+
+# For the mask data in the PNG file, the encoded raw pixel value need be to converted
+# to the proper class index. In the following map, [0, 0, 0] will be convert to 0, and
+# [128, 0, 0] will be conveted to 1, so on so forth. Also note that the mask class is 1
+# base since class 0 is reserved for the background. The [128, 0, 0] (class 1) is mapped
+# to `aeroplane`.
+VOC_PNG_COLOR_VALUE = [
+    [0, 0, 0],
+    [128, 0, 0],
+    [0, 128, 0],
+    [128, 128, 0],
+    [0, 0, 128],
+    [128, 0, 128],
+    [0, 128, 128],
+    [128, 128, 128],
+    [64, 0, 0],
+    [192, 0, 0],
+    [64, 128, 0],
+    [192, 128, 0],
+    [64, 0, 128],
+    [192, 0, 128],
+    [64, 128, 128],
+    [192, 128, 128],
+    [0, 64, 0],
+    [128, 64, 0],
+    [0, 192, 0],
+    [128, 192, 0],
+    [0, 64, 128],
+]
+VOC_PNG_COLOR_MAPPING = [0] * (256**3)
+for i, colormap in enumerate(VOC_PNG_COLOR_VALUE):
+    VOC_PNG_COLOR_MAPPING[(colormap[0] * 256 + colormap[1]) * 256 + colormap[2]] = i
+# There is a special mapping with [224, 224, 192] -> 255
+VOC_PNG_COLOR_MAPPING[224 * 256 * 256 + 224 * 256 + 192] = 255
+VOC_PNG_COLOR_MAPPING = tf.constant(VOC_PNG_COLOR_MAPPING)
```

Comment:
This variable occupies 64 MB, regardless if the dataset is being used or not.
I think this should be a sparse tensor, if possible, or at least lazy loaded:
```py
VOC_PNG_COLOR_MAPPING = None

def voc_png_color_mapping():
  global VOC_PNG_COLOR_MAPPING
  
  if VOC_PNG_COLOR_MAPPING is None:
    # generate...

  return VOC_PNG_COLOR_MAPPING
...

... tf.gather(voc_png_color_mapping(), ...)
```

Reply:
Good point. Change to lazy load to avoid unnecessary memory consumption when loading the python module.